src/frontend/src/binder/query.rs

```diff
@@ -289,51 +289,22 @@ impl Binder {
             let table_name = alias.name.real_value();
 
             if with.recursive {
-                let Query {
+                let (
+                    SetExpr::SetOperation {
+                        op: SetOperator::Union,
+                        all,
+                        left,
+                        right,
+                    },
                     with,
-                    body,
-                    order_by,
-                    limit,
-                    offset,
-                    fetch,
-                } = query;
-
-                /// the input clause should not be supported.
-                fn should_be_empty<T>(v: Option<T>, clause: &str) -> Result<()> {
-                    if v.is_some() {
-                        return Err(ErrorCode::BindError(format!(
-                            "`{clause}` is not supported in recursive CTE"
-                        ))
-                        .into());
-                    }
-                    Ok(())
-                }
-
-                should_be_empty(order_by.first(), "ORDER BY")?;
-                should_be_empty(limit, "LIMIT")?;
-                should_be_empty(offset, "OFFSET")?;
-                should_be_empty(fetch, "FETCH")?;
-
-                let SetExpr::SetOperation {
-                    op: SetOperator::Union,
-                    all,
-                    left,
-                    right,
-                } = body
+                ) = Self::validate_rcte(query)?
                 else {
                     return Err(ErrorCode::BindError(
-                        "`UNION` is required in recursive CTE".to_string(),
+                        "expect `SetOperation` as the return type of validation".into(),
```

Comment:
This is a user-facing error, while `SetOperation` is a uncommon terminology to users, so prefer keeping `UNION` here.

BTW, actually only `UNION` is valid here, other set operations will be rejected.

Reply:
OK, I got it. It's an internal error.

Reply:
yep and just for reference, the actual user-facing error is here: https://github.com/risingwavelabs/risingwave/pull/16458/files#diff-9defd28e358376355286b3fe37ff756a22c7bfd3bd5e25bb08848948dd2e09b5R373